cpp/src/lib/osd_access.cc

```diff
@@ -61,14 +61,38 @@ void OsdAccess::_remove_ctx(osd_t osd) {
   _osd_ctxs.erase(osd);
 }
 
-void OsdAccess::update(std::vector<std::pair<osd_t, info_caps>> &infos) {
-  ALBA_LOG(DEBUG, "OsdAccess::update");
-  std::lock_guard<std::mutex> lock(_osd_infos_mutex);
 
-  _osd_infos.clear();
-  for (auto &p : infos) {
-    _osd_infos.emplace(p.first, std::move(p.second));
-  }
+void OsdAccess::update(Proxy_client& client) {
+    if(!_filling){
```

Comment:
It's possible that 2 clients will enter this block simultaneously and still perform too much work.
This is typically solved by the following pattern (double checked locking):
```
if (need_to_do_sth) {
  take lock;
  if (need_to_do_sth) {
    do_thing();
    need_to_do_sth = false;
  } else {
    ...
  }
  release lock;
} else {
  ...
}